metadata/src/test/java/org/apache/kafka/metalog/LocalLogManager.java

```diff
@@ -729,52 +714,23 @@ public synchronized OptionalLong highWatermark() {
     }
 
     @Override
-    public long scheduleAppend(int epoch, List<ApiMessageAndVersion> batch) {
+    public long scheduleAppend(
+        int epoch,
+        List<ApiMessageAndVersion> batch
+    ) {
         if (batch.isEmpty()) {
             throw new IllegalArgumentException("Batch cannot be empty");
         }
 
-        List<ApiMessageAndVersion> first = batch.subList(0, batch.size() / 2);
-        List<ApiMessageAndVersion> second = batch.subList(batch.size() / 2, batch.size());
-
-        assertEquals(batch.size(), first.size() + second.size());
-        assertFalse(second.isEmpty());
-
-        OptionalLong firstOffset = first
-            .stream()
-            .mapToLong(record -> scheduleAtomicAppend(epoch,
-                    OptionalLong.empty(),
-                    Collections.singletonList(record)))
-            .max();
-
-        if (firstOffset.isPresent() && resignAfterNonAtomicCommit.getAndSet(false)) {
-            // Emulate losing leadership in the middle of a non-atomic append by not writing
-            // the rest of the batch and instead writing a leader change message
-            resign(leader.epoch());
-
-            return firstOffset.getAsLong() + second.size();
-        } else {
-            return second
-                .stream()
-                .mapToLong(record -> scheduleAtomicAppend(epoch,
-                        OptionalLong.empty(),
-                        Collections.singletonList(record)))
-                .max()
-                .getAsLong();
+        if (throwOnNextAppend.getAndSet(false)) {
+            throw new BufferAllocationException("Test asked to fail the next scheduleAppend");
         }
+
+        return shared.tryAppend(nodeId, leader.epoch(), batch);
```

Comment:
This isn't correct, since it means we are really appending the records immediately rather than waiting until `scheduleFlush`. You should instead just put the records into a temporary area and flush them in `scheduleFlush`.

Reply:
Yes. It is not correct but it is not trivial to implement. I tried to implement it but ran into many issues. I filed https://issues.apache.org/jira/browse/KAFKA-17339 to revisit this.